src/kv/kv_types.h

```diff
@@ -50,8 +50,58 @@ namespace kv
   DECLARE_JSON_TYPE(TxID);
   DECLARE_JSON_REQUIRED_FIELDS(TxID, term, version)
 
-  using BatchVector = std::vector<
-    std::tuple<kv::Version, std::shared_ptr<std::vector<uint8_t>>, bool>>;
+  // SeqNo indexes transactions processed by the consensus protocol providing
+  // ordering
+  using SeqNo = int64_t;
+
+  struct Configuration
+  {
+    struct NodeInfo
+    {
+      std::string hostname;
+      std::string port;
+      tls::Pem cert = {};
+
+      NodeInfo() = default;
+
+      NodeInfo(
+        const std::string& hostname_,
+        const std::string& port_,
+        const tls::Pem& cert_ = {}) :
+        hostname(hostname_),
+        port(port_),
+        cert(cert_)
+      {}
+    };
+
+    using Nodes = std::unordered_map<NodeId, NodeInfo>;
+
+    SeqNo idx;
+    Nodes nodes;
+  };
+
+  class ConfigurableConsensus
+  {
+  public:
+    virtual void add_configuration(
+      SeqNo seqno, const Configuration::Nodes& conf) = 0;
+    virtual Configuration::Nodes get_latest_configuration() const = 0;
```

Comment:
Looking at `add_configuration()` in `raft.h`, should `configurations` be protected by its own lock? I can see a few places where it is used, and could potentially be used by different threads simultaneously?

Reply:
That's currently assumed to be called under the Consensus lock (I'll add a comment in case it was not obvious). There's no real concurrency in the consensus itself right now, and I don't think making the config separately lockable is going to help that. Am I missing something obvious?

Reply:
I am going to change many things with locking with multi-threading on the backup so this seems fine for now?

Reply:
Yes, let's revisit locking if we think there's something to be gained from changing granularity. As I understand the changes you're making though, only execution will become concurrent, actual consensus logic remains sequential?